text/0001-zap-process.md

```diff
@@ -0,0 +1,106 @@
+implementation phase, where we work out a good design with an eye toward future
+maintainability before the details are set in stone.
+
+Some proposed changes may be controversial or involve deciding between a number
+of alternatives with different tradeoffs. These changes will benefit from
+discussion between all stakeholders, and a lightweight mechanism to record and
+collect the consensus decisions reached in those discussions, so that
+implementation can proceed in a consistent direction.
+
+Additionally, some proposed changes have a large enough scope that the
+implementation work is expected to be split over many pull requests over a long
+period of time. These changes will benefit from a detailed design document that
+guides the overall implementation plan and decreases the likelihood that earlier
+parts of the work need to be redone.
+
+# Detailed design
+
+When proposing a change that falls into the above categories or might otherwise
+benefit from a formalize design process, after preliminary discussion in our
+[development community](https://zulip.com/development-community/), a developer
+may choose to open a Zulip architecture proposal (ZAP).
+
+A ZAP is a pull request in this `zulip-architecture` repository that adds a
+Markdown document named like `text/0000-my-feature.md`. The numerical ID should
+be initialized with the placeholder `0000`, and replaced with the pull request
+number after the pull request is opened.
+
+The pull request description should link to the rendered Markdown document from
+the author's branch. The Markdown document should begin with a link back to the
+GitHub pull request page, and proceed with a suggested organization into
+sections such as “Summary”, “Motivation”, “Detailed design”, “Alternatives”,
+“Security implications”, “Future directions”, and “Previous discussion”, as
+applicable.
+
+Each new ZAP pull request should be posted to an appropriate stream in the
+development community for discussion. Discussion can happen in the development
+community or in comments/reviews on the pull request. Any discussion that
+happens outside the pull request should ideally be linked and summarized on the
+pull request to keep a record of it. The pull request will remain open while
+discussion is ongoing, and may be updated as desired.
+
+If the Zulip team reaches a consensus (via the usual decision-making conventions
+of the Zulip project, which are deliberately left unspecified here) that the
+proposal should be accepted, the pull request will be merged. If consensus is
+that the proposal should be rejected, the pull request will be closed, and if
+consensus is that it needs modifications, the proposal may remain open while
+being edited, or be closed in favor of a different pull request.
+
+The documents in `zulip-architecture` will be maintained to reflect the current
+consensus of the project. (This maintenance may involve using issues and pull
```

Comment:
This sounds a lot like it's saying that we'll maintain old accepted ZAPs to reflect the current design of the respective features.

I think that's not what you intend, because that was a major reason not to do this in the developer docs in the server/web repo.

What I imagine us doing is roughly:
* When we change our mind about something while an accepted ZAP is still in the process of being implemented, we'll typically edit the ZAP to reflect that.
* When the implementation is complete, we'll edit the metadata at the top of the ZAP to say so, and to link to the relevant subsystem doc and/or API docs.
* As we go on to tweak things thereafter, we generally won't try to update the ZAP to match. It remains a record of what we were thinking then; people finding the ZAP and interested in the up-to-date API are advised to follow the links to current docs.
* If we subsequently make a major overhaul — roughly, something that calls for a ZAP of its own — then we go back and add a link at the top of the older ZAP to point to the new one, a bit like the references at the top of old RFCs.

Reply:
I agree that ZAP should not be a living documentation but a document recording past consensus.

The proposed bullet points make sense to me. To implement this, we could use the concept of "status". A ZAP can be `draft` (when the initial PR is still open), `accepted` (when the initial PR is merged), or `final` (when the proposed change is implemented and documented).

Prior to the `final` phase, a ZAP can be modified, but major changes would require a new ZAP. A `final` ZAP should link to the relevant documentation in a `Final References` section. When the ZAP is superseded or overwritten, we link to the new ZAP in `Final References`.